File: Leetcode/Python/8. String to Integer (atoi)/solution.py
# https://leetcode.com/problems/string-to-integer-atoi/description/

class Solution:
    def myAtoi(self, s: str) -> int:
        # Delete space
        s = s.strip()

        if len(s) == 0:
            return 0

        idx = 0
        isNegative = False
        if s[0] in ["-", "+"]:
            if s[0] == "-":
                isNegative = True
            idx += 1

        # Read number from left
        number = 0
        while idx < len(s):
            if idx == len(s) or not s[idx].isdigit():
                break

            number *= 10
            number += int(s[idx])
            idx += 1

        if isNegative:
            number *= -1

        # Round number
        number = max(min(number, 2 ** 31 - 1), -2 ** 31)
        return number


class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.strip()   # Delete whitespace
        num = []
        isSign = False
        sign = 1

        for char in s:
            if char.isdigit():
                num.append(char)

            elif not num and char in ["+", "-"] and not isSign:
                if char == "-":
                    sign = -1
                isSign = True

            else:
                break

        if num:  # Rounding
            return max(min(int("".join(num)) * sign, 2 ** 31 - 1), -2 ** 31)

        return 0

File: Leetcode/Typescript/8. String to Integer (atoi)/solution.ts
// https://leetcode.com/problems/string-to-integer-atoi/description/

function myAtoi(s: string): number {
  s = s.trim();

  if (s.length === 0) {
    return 0;
  }

  let index = 0,
    isNegative = false;
  if (s[0] === "+" || s[0] === "-") {
    index += 1;
    if (s[0] === "-") {
      isNegative = true;
    }
  }

  // Read number from left
  let number = 0;
  while (index < s.length) {
    if (!isDigit(s[index])) {
      break;
    }
    number *= 10;
    number += parseInt(s[index]);
    console.log(number, s[index], Number.isNaN(s[index]));
    index += 1;
  }

  if (isNegative) {
    number *= -1;
  }
  // Round number
  number = Math.max(Math.min(number, 2 ** 31 - 1), (-2) ** 31);
  return number;
}

function isDigit(str: string): boolean {
  return /^[0-9]+$/.test(str);
}
